## Step 1: Import necessary libraries
Start by importing the required modules for data manipulation, modeling, and evaluation.

In [19]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

## **Step 2: Prepare your data**

For this example, you will create a synthetic dataset using NumPy. The X array will contain the feature (input) data, and the y array will contain the binary target (output) data (0s and 1s).

In [20]:
import kagglehub
import os

# 1. Download dataset terbaru (mengembalikan path folder)
path = kagglehub.dataset_download("dragonheir/logistic-regression")
#print("Path ke folder dataset:", path)

# 2. Cari nama file di dalam folder tersebut
files = os.listdir(path)
#print("File yang ditemukan:", files)

# 3. Gabungkan path folder dengan nama file
path = "/kaggle/input/logistic-regression"
os.listdir(path)
file_path = path + "/Social_Network_Ads.csv"

# 4. pisahkan training set dan testing set
df = pd.read_csv(file_path)
X = df[['Age', 'EstimatedSalary']]
y = df['Purchased']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Using Colab cache for faster access to the 'logistic-regression' dataset.


## **Step 3: Create and train the model**

Next, create an instance of the LogisticRegression class and train it using the `fit()` method on your training data.

In [21]:
# Create a logistic regression model instance
model = LogisticRegression(solver='liblinear', random_state=0)

# Train the model with the training data
model.fit(X_train, y_train)

LogisticRegression(random_state=0, solver='liblinear')

**Interpreting the coefficients:**

After training, you can inspect the model's coefficients `(coef_)` and intercept `(intercept_)`. The coefficient represents the change in the log-odds of the outcome for a one-unit change in the predictor variable.


In [22]:
# Get the model's coefficients and intercept
print(f"Intercept: {model.intercept_}")
print(f"Coefficients: {model.coef_}")

# You can convert the coefficients to odds ratios by exponentiating them
odds_ratio = np.exp(model.coef_)
print(f"Odds Ratio: {odds_ratio}")
# An odds ratio of 4.03, for instance, means that for every 1mm increase in tumor size,
# the odds of it being cancerous increase by 4x.

Intercept: [-0.96624957]
Coefficients: [[2.06240531 1.05346464]]
Odds Ratio: [[7.86486453 2.86756901]]


## **Step 4: Make predictions**

Use the trained model to make predictions on the test data. The `predict()` method returns the predicted class (0 or 1), while `predict_proba()` returns the probabilities of each class.

In [23]:
# Predict the class labels for the test set
y_pred = model.predict(X_test)
print(f"Predicted labels: {y_pred}")

# Predict the probability of each class
y_pred_proba = model.predict_proba(X_test)
print(f"Predicted probabilities:\n {y_pred_proba}")

Predicted labels: [0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 1 0 1 0 1 0 0 0 0 0 0 1 0 0 0 0
 0 0 1 0 0 0 0 1 0 0 1 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 0 0 1 0 0 0 0 1 0 0 0
 0 0 1 0 1 1 1 1 0 0 1 1 0 1 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 1 1 0 0 0 0 0
 0 0 1 1 1 1 0 1 1]
Predicted probabilities:
 [[0.88417837 0.11582163]
 [0.82378966 0.17621034]
 [0.80014495 0.19985505]
 [0.90689755 0.09310245]
 [0.89555943 0.10444057]
 [0.99080965 0.00919035]
 [0.98243116 0.01756884]
 [0.28555006 0.71444994]
 [0.99355395 0.00644605]
 [0.4839337  0.5160663 ]
 [0.96046703 0.03953297]
 [0.96807405 0.03192595]
 [0.83091258 0.16908742]
 [0.62655804 0.37344196]
 [0.98345603 0.01654397]
 [0.67134089 0.32865911]
 [0.71542389 0.28457611]
 [0.98684504 0.01315496]
 [0.01296641 0.98703359]
 [0.95265789 0.04734211]
 [0.9038041  0.0961959 ]
 [0.04143657 0.95856343]
 [0.71201245 0.28798755]
 [0.12421254 0.87578746]
 [0.99475613 0.00524387]
 [0.03064172 0.96935828]
 [0.9196058  0.0803942 ]
 [0.92041959 0.07958041]
 [0.7988066

## **Step 5: Evaluate the model**

Evaluate your model's performance using metrics like accuracy, a confusion matrix, and a classification report.

In [24]:
# Calculate the accuracy of the model on the test set
accuracy = model.score(X_test, y_test)
print(f"Model accuracy: {accuracy:.2f}")

# Display the confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)
# The confusion matrix shows:
# [True Negatives, False Positives]
# [False Negatives, True Positives]

# Display a more comprehensive classification report
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Model accuracy: 0.88
Confusion Matrix:
[[74  5]
 [10 31]]
Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.94      0.91        79
           1       0.86      0.76      0.81        41

    accuracy                           0.88       120
   macro avg       0.87      0.85      0.86       120
weighted avg       0.87      0.88      0.87       120



## **Analisis yang Diminta**
***
###1. Bagaimana kinerja model berdasarkan classification report?
Kinerja model tergolong baik dengan nilai akurasi keseluruhan sebesar 88%, Namun jika dilihat lebih detail per kelas:

- Kelas 0 (Tidak Membeli): Memiliki performa yang sangat kuat dengan F1-score 0.91 Nilai recall 0.94 menunjukkan model sangat mahir mengenali orang yang memang tidak berniat membeli

- Kelas 1 (Membeli): Memiliki precision 0.86, yang berarti dari semua orang yang diprediksi akan membeli, 86% benar-benar melakukan pembelian Namun, recall-nya sebesar 0.76 menunjukkan ada sekitar 24% calon pembeli aktual yang gagal dideteksi oleh model (terlewat).

- Keseimbangan: Nilai weighted average untuk F1-score (0.87) menunjukkan model cukup stabil dalam menangani kedua kelas, meskipun performa pada kelas mayoritas (0) sedikit lebih unggul karena jumlah datanya lebih banyak (support 79 vs 41)
***

###2. Apa makna True Positive, False Positive, True Negative, dan False Negative padastudi kasus yang Anda pilih?

Dalam konteks prediksi pembelian iklan, berikut adalah maknanya =

- True Positive (31): Model memprediksi pengguna akan membeli, dan faktanya pengguna tersebut benar-benar membeli. Ini adalah target promosi yang tepat sasaran.

- False Positive (5): Model memprediksi pengguna akan membeli, padahal nyatanya pengguna tersebut tidak membeli Dalam bisnis, ini berarti "pemborosan biaya iklan" karena mengirim promosi ke orang yang tidak tertarik.

- True Negative (74): Model memprediksi pengguna tidak akan membeli, dan faktanya pengguna tersebut memang tidak membeli.

- False Negative (10): Model memprediksi pengguna tidak akan membeli, padahal pengguna tersebut sebenarnya ingin membeli,Ini adalah "kehilangan peluang" (missed opportunity) bagi perusahaan.
***
###3. Berikan kesimpulan mengenai seberapa baik Logistic Regression bekerja pada dataset yang digunakan.

Secara keseluruhan, Logistic Regression bekerja sangat baik pada dataset ini dengan akurasi 0.88.
Model ini sangat efektif sebagai model dasar (baseline) karena sifatnya yang sederhana dan mudah diinterpretasikan. Logistic Regression berhasil menangkap hubungan antara fitur (seperti umur dan gaji) dengan keputusan pembelian secara efisien.Meskipun terdapat beberapa kesalahan klasifikasi (terutama pada False Negative), performanya sudah cukup stabil untuk digunakan dalam analisis profil konsumen guna mengotomatisasi strategi pemasaran.